Macos ci

yedayak opened the pull request:
See https://github.com/scop/bash-completion/issues/726

Cherry-picked from here:
- [x] https://github.com/scop/bash-completion/pull/1440
- [x] https://github.com/scop/bash-completion/pull/1441
- [x] https://github.com/scop/bash-completion/pull/1442
- [x] https://github.com/scop/bash-completion/pull/1456

scop committed:
test: add macOS test script

scop committed:
ci(macos): add macOS CI test config

yedayak committed:
ci(macos): Align macos script with current linux

Also use venv instead of global install
brew/python doesn't like installing python packages globally:
```
error: externally-managed-environment
× This environment is externally managed
```

yedayak commented:
The CI run failed because for some reason it didn't manage to delete the temp directory it creates in `test_unit_load.py` using shutil.rmtree in `prepare_fixture_dir`

yedayak committed:
ci(macos): Force colored output

scop commented:
Thanks for working on this! I think I've seen the temp dir removal issue myself as well. Would be great if we could figure out a workaround. If we can't, we should allow it to fail only on macOS.

yedayak committed:
test(vipw): Drop exception for macos

vipw seems to have -d as an option on macos now.
Partially reverts ccf7bf69d "test(dmesg,vipw): expect no options on macOS"

yedayak committed:
test: Sort completion results items

It seems there is some behaviour differences regarding the order
completions are returned in between linux and macos, specifically
regarding sorting of upper case and lowercase letters. Sorting them
ourselves in the tests makes it consistent.

yedayak committed:
test: Remove empty strings from completion results

yedayak committed:
test(tar): skip test_25 for non-gnu tar

scop commented on test/t/unit/test_unit_dequote.py:
```diff
@@ -38,7 +40,8 @@ def test_5_brace(self, bash, functions):
     def test_6_glob(self, bash, functions):
         output = assert_bash_exec(bash, "__tester 'a?b'", want_output=True)
-        assert output.strip() == "<a b><a$b><a&b><a'b>"
+        items = sorted(re.findall(r"<[^>]*>", output))
```
It's not immediately clear to me why this is needed on macOS and not on others, could we elaborate in a comment?

yedayak replied:
Yeah I'm not sure what's up with that. I just checked and LC_COLLATE seems to be set to C. I'll need to look into that

yedayak committed:
Revert "test: Sort completion results items"

This reverts commit 8074741e7a7830a295f0c1ddbfd4115d3a0f7f3d.

yedayak commented on test/t/unit/test_unit_load.py:
```diff
@@ -33,7 +34,11 @@ def fixture_dir(self, request, bash):
-        return str(tmpdir)
+        yield str(tmpdir)
+        if sys.platform == "darwin":
+            assert_bash_exec(bash, "sudo rm -rf %s/*" % tmpdir)
```
> I wonder why we need to do this manual removal in the first place. In prepare_fixture_dir we add a finalizer that is supposed to clean up the temp dir altogether. Does that not work here? Does it work in other places where we use it to set up a fixture dir?

From what I gathered, the issue is that this is the only place where we create more files and folders inside the fixture_dir. Those extra files are created by other processes and not python, so I think that's what causes the issue. See [this note](https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man3/removefile.3.html#NOTES:~:text=Write%20protected%20files%20owned%20by%20another%20process%20cannot%20be%20removed%20by%20removefile%28%29%2C%20regardless%20of%20the%20permissions%20on%20the%20directory%20containing%20the%20file%2E) in the `removefile()` syscall: 
>> Write protected files owned by another process cannot be removed by
>> removefile(), regardless of the permissions on the directory containing
>> the file.

I might have just not found the correct magic spell to make these files deletable without sudo

> 
> But if it is needed, some followup questions:
> 
> Can we assume sudo is present? What if it asks for credentials?

Yeah that's a problem. It seems the default behavior is asking [for a password](https://support.apple.com/en-il/guide/terminal/apd5b0b6259-a7d4-4435-947d-0dff528912ba/mac#:~:text=You%E2%80%99re%20asked%20for%20the%20password%20of%20the%20current%20user)
> 
> Any idea if this is needed on all macOS setups, or just the GH one?

I don't know, the man pages seem to imply this will happen for all macOS's, but maybe that's not the real issue. I don't have a physical one to test on.

> If it's a GH only thing, and perhaps otherwise as well, I suppose we could just not fail on error here if the remove fails and let the error message come through so people can clean up manually.

I'm not sure what a good way to let the error come through without failing the tests is
> 
> Let's add a comment with details why it doesn't work without it here.
> 
> Maybe consider doing the sudo rm in the prepare_fixture_dir finalizer on darwin, if it is needed?

`prepare_fixture_dir` doesn't currently have access to bash, so it can't run sudo. It also isn't needed for all of the uses of it. I can make it accept a `bash` instance if you think that's better though.

yedayak committed:
test sorting with LC_COLLATE

scop commented on .github/workflows/ci.yaml:
```diff
@@ -103,3 +103,11 @@ jobs:
         if: steps.release.outputs.release_created
+
+  distcheck-macos:
+    runs-on: macos-latest
```
macos-latest will switch to 26 in coming weeks, https://github.com/actions/runner-images/issues/14167

Hopefully that'll help rather than hurt some :)

yedayak replied:
Tried now, didn't change. The only issue really left is the sorting of completion items being different, which I think is done by readline using `strcoll(3)` if available

yedayak committed:
ci(macos): Try macos-26
